### PR TITLE
Disable `skip_failed_rules` optimizer config by default

### LINKED ISSUE #4615: Don't ignore failed optimizer rules
**Describe the bug**
Some logical optimizer rules like type coercion are actually essential for correct query execution, but are currently silently ignored when they failed.

**To Reproduce**
TBD

**Expected behavior**
Don't skip failed rules (this probably requires us to remove the non-failing `OptimizerRule::optimize` method).

**Workaround**
This behavior is controlled by a config option (that defaults to `true` meaning to silently skip failed errors). 

You can see this in the datafusion cli:

```sql
DataFusion CLI v20.0.0
❯ show all;
+-----------------------------------------------------------+------------+
| name                                                      | setting    |
+-----------------------------------------------------------+------------+
...
| datafusion.optimizer.skip_failed_rules                    | true       |
...
+-----------------------------------------------------------+------------+
35 rows in set. Query took 0.038 seconds.
```

Datafusion can be configured to fail hard if an optimizer rule fails like

```sql
❯ -- default behavior
❯ set datafusion.optimizer.skip_failed_rules = false;
0 rows in set. Query took 0.023 seconds.
```

**Additional context**
See [this discussion](https://github.com/apache/arrow-datafusion/pull/4599#discussion_r1048288118).

### PR DESCRIPTION
**Pull request by jackwener**
# Which issue does this PR close?

Closes #4615.

# Rationale for this change

#4685 

# What changes are included in this PR?

Disable skip_failed_rules default

# Are these changes tested?

# Are there any user-facing changes?

YES!!!
This PR change default behavior.

### CODE REVIEW

**Review by alamb — APPROVED** [review]
Thank you @jackwener  ❤️

File: datafusion/core/tests/sql/subqueries.rs [hunk 1]
```diff
@@ -576,6 +576,8 @@ async fn support_order_by_correlated_columns() -> Result<()> {
     Ok(())
 }
 
+// TODO: issue https://github.com/apache/arrow-datafusion/issues/6263
```
Comment (comphead) [inline datafusion/core/tests/sql/subqueries.rs:579]: 👍 so only 1 test gets failed on optimization rules?

Comment (jackwener) [reply datafusion/core/tests/sql/subqueries.rs:579]: Yes

Comment (mingmwang) [reply datafusion/core/tests/sql/subqueries.rs:579]: Which rule fail this test ?

Comment (jackwener) [reply datafusion/core/tests/sql/subqueries.rs:579]: > Which rule fail this test ?

decorrelate_where_in, you can see issue #6263

Comment (mingmwang) [reply datafusion/core/tests/sql/subqueries.rs:579]: I will fix this rule later.

**Pull request merged**